sysvabi64/sysvabi64.rst

```diff
@@ -916,10 +923,639 @@ These follow the AAELF64_ base definition from Arm.
+when BTI compliant PLT entries requiring signed ``.got.plt[N]``
+entries are used.
+
+PLT header ``.got.plt[0]``
+
+.. code-block:: asm
+
+    bti  c
+    stp  x16, x30, [sp,#-16]!
+    adrp x16, Page(G(.plt.got[2]))
+    ldr  x17, [x16, G(.plt.got[2]) - GOTPLT]
+    add  x16, x16, G(.plt.got[2]) - GOTPLT
+    br   x17
+
+PLT entry ``.got.plt[N]``
+
+.. code-block:: asm
+
+    bti  c
+    adrp x16, Page(G(.plt.got[N]))
+    ldr  x17, [x16, G(.plt.got[N]) - GOTPLT]
+    add  x16, x16, G(.plt.got[N]) - GOTPLT
+    autia1716
+    br   x17
+
+GNU Indirect Functions
+----------------------
+
+A GNU Indirect Function (IFUNC) is a feature that permits a single
+implementation of a function to be chosen from multiple candidates,
+with the choice taken by an IFUNC resolver function.
+
+GNU Indirect Functions require static and dynamic linker support. They
+are known to be supported on GNU/Linux, Android and many of the BSD
+operating systems.
+
+GNU Indirect Functions are called via a PLT entry that loads the
+function address. The function address is chosen by an IFUNC resolver
+function.
+
+The source code interface to an IFUNC resolver is platform
+dependent. The GNU/Linux interface via the GNU C Library is documented
+below.
+
+GNU C Library IFUNC interface
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+The GNU C Library (glibc) ``sys/ifunc.h`` header provides the
+necessary type definitions for defining IFUNC resolvers. The resolver
+is passed at least one, and at most 2 parameters dependent on the
+glibc IFUNC ABI version. A flag is set in the first parameter if a
+second parameter is available. The parameters provide access to the
+HWCAP_ flags.
+
+.. code-block:: c
+
+  /* A second argument is passed to the ifunc resolver.  */
+  #define _IFUNC_ARG_HWCAP        (1ULL << 62)
+
+  /* Function prototype for an IFUNC resolver.
+  ElfW(Addr) ifunc_resolver (uint64_t, const __ifunc_arg_t *);
+
+  /* Second argument to an ifunc resolver.  */
+  struct __ifunc_arg_t
+  {
+    unsigned long _size; /* Size of the struct, so it can grow.  */
+    unsigned long _hwcap;
+    unsigned long _hwcap2;
+  };
+
+  typedef struct __ifunc_arg_t __ifunc_arg_t;
+
+IFUNC resolver functions must have a type of ``STT_GNU_IFUNC``. With
+the GCC and Clang compilers an attribute can be used to achieve this.
+
+.. code-block:: c
+
+  /* Implementation alternative 1 */
+  static int implementation1(void) { ... }
+
+  /* Implementation alternative 2 */
+  static int implementation2(void) { ... }
+
+  /* Resolver function */
+  static void* resolver(uint64_t, const __ifunc_arg_t *) { ... }
+
+  /* make symbol ifunc type STT_GNU_IFUNC using resolver as the IFUNC resolver. */
+  int ifunc(void) __attribute__((ifunc("resolver")));
+
+The first ``uint64`` parameter is ``AT_HWCAP``. If the
+``_IFUNC_ARG_HWCAP`` flag is set the second parameter is present. The
+second parameter is a structure ``__ifunc_arg_t`` defined in
+``sys/ifunc.h`` it provides access to all of the flags in HWCAP_ via
```

Comment:
Omitted punctuation before `it provides ...`

Reply:
Ack.